rust/src/lexer.rs

```diff
@@ -61,60 +63,66 @@ impl<'a, 'b> Lexer<'a, 'b> {
   fn is_symbol_ident(ch: char) -> bool {
     !(ch.is_whitespace() || ch.is_alphanumeric() || "[]{};,".contains(ch) || "()\"\'`".contains(ch))
   }
-}
-impl<'a, 'b> Iterator for Lexer<'a, 'b> {
-  type Item = Result<Token<'a>, SourceError<'a, LexerError>>;
 
-  fn next(&mut self) -> Option<Self::Item> {
-    let mut ch;
-    let mut position;
-    loop {
-      ch = self.iterator.next()?;
-      position = self.iterator.position.previous_column();
-      if ch.is_whitespace() {
-        // do nothing
-      } else if ch == '/' {
-        if self.iterator.peek() == Some('/') {
+  fn next_equals(&mut self, other: char) -> Option<bool> {
+    return self.iterator.peek().map(|x| x == other);
+  }
+
+  fn lex_doc(&mut self) -> Option<LexItem<'a>> {
+    if self.next_equals('/')? {
+      self.iterator.next();
+      if self.next_equals('/')? {
+        self.iterator.next();
+        // we are in a comment, or a doc-string
+        if self.next_equals('/')? {
           self.iterator.next();
-          // we are in a comment, or a doc-string
-          if self.iterator.peek() == Some('/') {
-            self.iterator.next();
-            // it's a doc-string
-            let doc = self.iterator.until_eol();
-            let tok = Token {
-              position,
-              symbol: Symbol::Doc(doc),
-            };
-            return Some(Ok(tok));
-          } else {
-            // it's a comment
-            self.iterator.until_eol();
-          }
+          let position = self.iterator.position;
+          // it's a doc-string
+          let doc = self.iterator.until_eol();
+          let tok = Token {
+            position,
+            symbol: Symbol::Doc(doc),
+          };
+          return Some(Ok(tok));
         } else {
-          break;
+          // it's a comment
+          self.iterator.until_eol();
+          None
         }
       } else {
-        break;
+        None
       }
+    } else {
+      None
     }
-    let mut negate = false;
-    if ch == '-' && self.iterator.peek().map_or(false, |x| x.is_ascii_digit()) {
-      negate = true;
-      ch = self.iterator.next()?;
-      position = self.iterator.position.previous_column();
-    }
-    if ch.is_ascii_digit() {
+  }
+  fn lex_int_lit(&mut self) -> Option<LexItem<'a>> {
+    let negation = if self.next_equals('-')? {
+      // Skip the peeked dash.
+      self.iterator.next()?;
+      Some(self.iterator.position.previous_column())
```

Comment:
Note that we must not parse subtractions as negations.. e.g. 3 - 4 must still parse OK. That's why I require in the original code that the - must be followed immediately by a digit for negation to apply.

Reply:
All right, I'll add tests for subtraction and negation and make sure it works as expected.